### PR TITLE
[MD][Cred mngment] - PoC Create credential

### PR DESCRIPTION
**Pull request by noCharger**
### Description
 1. Route to creation page from list page
 2. Create credential with password encrypted

Screen Shot 2022-07-07 at 9 44 38 AM
Screen Shot 2022-07-07 at 9 44 13 AM

### CODE REVIEW

File: src/plugins/data_source_management/public/components/create_data_source_wizard/components/header/header.tsx [hunk 1]
```diff
@@ -27,8 +27,7 @@ export const Header = ({
     {
       defaultMessage: 'Create {dataSourceName}',
       values: { dataSourceName },
-    }
-  );
+    });
```
Comment (seraphjiang) [inline src/plugins/data_source_management/public/components/create_data_source_wizard/components/header/header.tsx:30]: thanks

File: src/plugins/credential_management/server/routes/create.ts [hunk 2]
```diff
@@ -24,11 +24,11 @@ export function registerCreateRoute(router: IRouter) {
                   schema.literal('basic_auth'),
                   schema.literal('aws_iam_credential'),
                 ]),
-                basic_auth_credential_JSON: schema.maybe(schema.object({
+                basic_auth_credential_json: schema.maybe(schema.object({
```
Comment (seraphjiang) [inline src/plugins/credential_management/server/routes/create.ts:27]: weird name to me,
how about name it `username_password_credential`
no need to add json postfix

File: src/plugins/credential_management/public/components/create_credential_wizard/components/header/header.tsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            </>
+          ) : null}
+        </h1>
+      </EuiTitle>
+      <EuiSpacer size="s" />
+      <EuiText>
+        <p>
+          <FormattedMessage
+            id="credentialManagement.createCredential.description"
+            defaultMessage="An credential can be attached to multiple sources. For example, {credential} can be attached to two data sources {first} and {second}."
```
Comment (seraphjiang) [inline src/plugins/credential_management/public/components/create_credential_wizard/components/header/header.tsx:61]: typo `An credential`

File: src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx [hunk 4]
```diff
@@ -8,3 +8,204 @@
...
+  public readonly context!: CredentialManagmentContextValue;
+
+  constructor(props: RouteComponentProps, context: CredentialManagmentContextValue) {
+    super(props, context);
+
+    context.services.setBreadcrumbs(getCreateBreadcrumbs());
+
+    this.state = {
+      credentialName: '',
+      credentialType: 'basic_auth',
```
Comment (seraphjiang) [inline src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx:57]: prefer `username_password_credential`, and make it const

File: src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx [hunk 5]
```diff
@@ -8,3 +8,204 @@
...
+              user_name: this.state.user_name,
+              password: this.state.password,
+            },
+          }),
+        })
+        .then((res) => {
+          console.info(res);
+        });
+    } catch (e) {
+      return e;
```
Comment (seraphjiang) [inline src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx:205]: could we log the error?

and why return error?

File: src/plugins/credential_management/public/service/creation/manager.ts [hunk 6]
```diff
@@ -10,61 +10,59 @@
...
-    //     return getType('default');
-    //   }
-    // };
+        if (config) {
+          return config;
+        } else {
+          throw new Error(`Credential creation type not found: ${key}`);
+        }
+      } else {
+        return getType('default');
```
Comment (seraphjiang) [inline src/plugins/credential_management/public/service/creation/manager.ts:44]: what if on entry for `default` will this cause infinity recursive call?

**Review by seraphjiang — APPROVED** [review]

File: src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx [hunk 7]
```diff
@@ -8,3 +8,204 @@
...
+            <EuiFormRow label="Password">
+              <EuiFieldText
+                placeholder="Your Password"
+                value={this.redact(this.state.password.length) || ''}
+                onChange={(e) => this.setState({ password: e.target.value })}
+              />
+            </EuiFormRow>
+            <EuiFormRow label="Upload Credential File">
+              <EuiFilePicker />
+            </EuiFormRow>
```
[the remark this answers is not in the document]
Comment (noCharger) [reply src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx:150]: It's just a placeholder where user could potentially submit credential file

File: src/plugins/credential_management/public/plugin.ts [hunk 8]
```diff
@@ -34,23 +32,52 @@ export interface CredentialManagementStartDependencies {
...
+    urlForwarding.forwardApp(
+      'management/opensearch-dashboards/credentials',
+      newAppPath,
+      (path) => '/create'
+    );
+
+    urlForwarding.forwardApp(legacyPatternsPath, newAppPath, (path) => {
+      const pathInApp = path.substr(legacyPatternsPath.length + 1);
+      return pathInApp && `/patterns${pathInApp}`;
+    });
```
[the remark this answers is not in the document]
Comment (noCharger) [reply src/plugins/credential_management/public/plugin.ts:76]: It is used for Dock navigation:

Screen Shot 2022-07-08 at 7 58 45 PM

File: src/plugins/credential_management/public/service/creation/manager.ts (hunk 6 above)
Comment (noCharger) [reply src/plugins/credential_management/public/service/creation/manager.ts:44]: removed the getType logic

**Comment by seraphjiang** [comment]
try save a credential, but got error?

i guess it is because missing crypto_material?

if so, how could i got a working dummy crypto_material file?

```
server    log   [00:56:48.119] [info][plugins-system] Stopping all plugins.
 restarting server  due to changes in "src/plugins/credential_management/server/crypto/crypto_material"
server    log   [00:56:51.422] [info][plugins-service] Plugin "visTypeXy" is disabled.
server    log   [00:56:51.514] [info][plugins-system] Setting up [38] plugins: [usageCollection,opensearchDashboardsUsageCollection,opensearchDashboardsLegacy,mapsLegacy,share,opensearchUiShared,legacyExport,embeddable,expressions,data,home,console,apmOss,management,indexPatternManagement,dataSourceManagement,advancedSettings,savedObjects,dashboard,visualizations,visTypeVega,visTypeTimeline,timeline,visTypeTable,visTypeMarkdown,tileMap,regionMap,inputControlVis,visualize,credentialManagement,charts,visTypeVislib,visTypeTimeseries,visTypeTagcloud,visTypeMetric,discover,savedObjectsManagement,bfetch]
server    log   [00:56:51.722] [info][savedobjects-service] Waiting until all OpenSearch nodes are compatible with OpenSearch Dashboards before starting saved objects migrations...
server    log   [00:56:51.743] [info][savedobjects-service] Starting saved objects migrations
server    log   [00:56:51.787] [info][plugins-system] Starting [38] plugins: [usageCollection,opensearchDashboardsUsageCollection,opensearchDashboardsLegacy,mapsLegacy,share,opensearchUiShared,legacyExport,embeddable,expressions,data,home,console,apmOss,management,indexPatternManagement,dataSourceManagement,advancedSettings,savedObjects,dashboard,visualizations,visTypeVega,visTypeTimeline,timeline,visTypeTable,visTypeMarkdown,tileMap,regionMap,inputControlVis,visualize,credentialManagement,charts,visTypeVislib,visTypeTimeseries,visTypeTagcloud,visTypeMetric,discover,savedObjectsManagement,bfetch]
server    log   [00:56:51.972] [info][listening] Server running at http://localhost:5601
server    log   [00:56:52.006] [info][server][OpenSearchDashboards][http] http server running at http://localhost:5601
server    log   [00:57:04.225] [error][http] SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at new CryptoCli (/Users/huanji/src/ea-osd/src/plugins/credential_management/server/crypto/cli/crypto_cli.ts:35:34)
    at Function.getInstance (/Users/huanji/src/ea-osd/src/plugins/credential_management/server/crypto/cli/crypto_cli.ts:66:29)
    at createHandler (/Users/huanji/src/ea-osd/src/plugins/credential_management/server/credential/credential_manager.ts:21:31)
    at handler (/Users/huanji/src/ea-osd/src/plugins/credential_management/server/routes/create.ts:44:32)
    at /Users/huanji/src/ea-osd/src/core/utils/context.ts:259:14
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at Router.handle (/Users/huanji/src/ea-osd/src/core/server/http/router/router.ts:286:44)
    at handler (/Users/huanji/src/ea-osd/src/core/server/http/router/router.ts:241:11)
    at exports.Manager.execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/toolkit.js:60:28)
    at Object.internals.handler (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/handler.js:46:20)
    at exports.execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/handler.js:31:20)
    at Request._lifecycle (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/request.js:371:32)
    at Request._execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/request.js:281:9)
server   error  [00:57:04.221]  Error: Internal Server Error
    at HapiResponseAdapter.toInternalError (/Users/huanji/src/ea-osd/src/core/server/http/router/response_adapter.ts:71:19)
    at Router.handle (/Users/huanji/src/ea-osd/src/core/server/http/router/router.ts:303:34)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at handler (/Users/huanji/src/ea-osd/src/core/server/http/router/router.ts:241:11)
    at exports.Manager.execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/toolkit.js:60:28)
    at Object.internals.handler (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/handler.js:46:20)
    at exports.execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/handler.js:31:20)
    at Request._lifecycle (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/request.js:371:32)
    at Request._execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/request.js:281:9)
```
image

**Comment by seraphjiang** [comment]
don't quite understand what does credential file do? why do we need this?

image

**Comment by noCharger** [comment]
> try save a credential, but got error?
> 
> i guess it is because missing crypto_material?
> 
> if so, how could i got a working dummy crypto_material file?
> 
> ```
> server    log   [00:56:48.119] [info][plugins-system] Stopping all plugins.
>  restarting server  due to changes in "src/plugins/credential_management/server/crypto/crypto_material"
> server    log   [00:56:51.422] [info][plugins-service] Plugin "visTypeXy" is disabled.
> server    log   [00:56:51.514] [info][plugins-system] Setting up [38] plugins: [usageCollection,opensearchDashboardsUsageCollection,opensearchDashboardsLegacy,mapsLegacy,share,opensearchUiShared,legacyExport,embeddable,expressions,data,home,console,apmOss,management,indexPatternManagement,dataSourceManagement,advancedSettings,savedObjects,dashboard,visualizations,visTypeVega,visTypeTimeline,timeline,visTypeTable,visTypeMarkdown,tileMap,regionMap,inputControlVis,visualize,credentialManagement,charts,visTypeVislib,visTypeTimeseries,visTypeTagcloud,visTypeMetric,discover,savedObjectsManagement,bfetch]
> server    log   [00:56:51.722] [info][savedobjects-service] Waiting until all OpenSearch nodes are compatible with OpenSearch Dashboards before starting saved objects migrations...
> server    log   [00:56:51.743] [info][savedobjects-service] Starting saved objects migrations
> server    log   [00:56:51.787] [info][plugins-system] Starting [38] plugins: [usageCollection,opensearchDashboardsUsageCollection,opensearchDashboardsLegacy,mapsLegacy,share,opensearchUiShared,legacyExport,embeddable,expressions,data,home,console,apmOss,management,indexPatternManagement,dataSourceManagement,advancedSettings,savedObjects,dashboard,visualizations,visTypeVega,visTypeTimeline,timeline,visTypeTable,visTypeMarkdown,tileMap,regionMap,inputControlVis,visualize,credentialManagement,charts,visTypeVislib,visTypeTimeseries,visTypeTagcloud,visTypeMetric,discover,savedObjectsManagement,bfetch]
> server    log   [00:56:51.972] [info][listening] Server running at http://localhost:5601
> server    log   [00:56:52.006] [info][server][OpenSearchDashboards][http] http server running at http://localhost:5601
> server    log   [00:57:04.225] [error][http] SyntaxError: Unexpected end of JSON input
>     at JSON.parse (<anonymous>)
>     at new CryptoCli (/Users/huanji/src/ea-osd/src/plugins/credential_management/server/crypto/cli/crypto_cli.ts:35:34)
>     at Function.getInstance (/Users/huanji/src/ea-osd/src/plugins/credential_management/server/crypto/cli/crypto_cli.ts:66:29)
>     at createHandler (/Users/huanji/src/ea-osd/src/plugins/credential_management/server/credential/credential_manager.ts:21:31)
>     at handler (/Users/huanji/src/ea-osd/src/plugins/credential_management/server/routes/create.ts:44:32)
>     at /Users/huanji/src/ea-osd/src/core/utils/context.ts:259:14
>     at processTicksAndRejections (internal/process/task_queues.js:95:5)
>     at Router.handle (/Users/huanji/src/ea-osd/src/core/server/http/router/router.ts:286:44)
>     at handler (/Users/huanji/src/ea-osd/src/core/server/http/router/router.ts:241:11)
>     at exports.Manager.execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/toolkit.js:60:28)
>     at Object.internals.handler (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/handler.js:46:20)
>     at exports.execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/handler.js:31:20)
>     at Request._lifecycle (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/request.js:371:32)
>     at Request._execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/request.js:281:9)
> server   error  [00:57:04.221]  Error: Internal Server Error
>     at HapiResponseAdapter.toInternalError (/Users/huanji/src/ea-osd/src/core/server/http/router/response_adapter.ts:71:19)
>     at Router.handle (/Users/huanji/src/ea-osd/src/core/server/http/router/router.ts:303:34)
>     at processTicksAndRejections (internal/process/task_queues.js:95:5)
>     at handler (/Users/huanji/src/ea-osd/src/core/server/http/router/router.ts:241:11)
>     at exports.Manager.execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/toolkit.js:60:28)
>     at Object.internals.handler (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/handler.js:46:20)
>     at exports.execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/handler.js:31:20)
>     at Request._lifecycle (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/request.js:371:32)
>     at Request._execute (/Users/huanji/src/ea-osd/node_modules/@hapi/hapi/lib/request.js:281:9)
> ```
> 
> image

Check readme - https://github.com/elastic-analytics/OpenSearch-Dashboards/tree/2.x/src/plugins/credential_management

**Comment by noCharger** [comment]
> don't quite understand what does credential file do? why do we need this?
> 
> image

It's placeholder allowing user to submit their own crypto_material.

**Comment by seraphjiang** [comment]
> 

good idea, but i'm not sure if there is industry standard for such kind of file type.

**Comment by seraphjiang** [comment]
> 

Thanks, readme helped.

there are a few ux issues which we could work with ux to find better experience.

1. we should call `Save credential` (not creation. we don't create any, it exist other place, we just enter and save )
2. after saved, it is better to redirect to list page

**Comment by noCharger** [comment]
> > 
> 
> good idea, but i'm not sure if there is industry standard for such kind of file type.

Because it is not a top priority, I would remove this UI component for the time being and follow up later: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1872

**Comment by noCharger** [comment]
> > 
> 
> Thanks, readme helped.
> 
> there are a few ux issues which we could work with ux to find better experience.
> 
> 1. we should call `Save credential` (not creation. we don't create any, it exist other place, we just enter and save )
> 2. after saved, it is better to redirect to list page

Thanks for the feedback, should included in the next PR.

File: src/plugins/credential_management/common/types.ts [hunk 9]
```diff
@@ -9,19 +9,25 @@
...
+export type AWS_IAM_TYPE = 'aws_iam_credential';
+
+export interface IBasicAuthCredentialMaterial {
   readonly user_name: string;
   readonly password: string;
 }
 
-interface IAWSIAMCredentialMaterial {
+export interface IAWSIAMCredentialMaterial {
   readonly encrypted_aws_iam_credential: string;
```
Comment (zengyan-amazon) [inline src/plugins/credential_management/common/types.ts:32]: how do we use this `encrypted_aws_iam_credential`?

File: src/plugins/credential_management/public/components/breadsrumbs.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  return [
+    ...getListBreadcrumbs(),
+    {
+      text: i18n.translate('credentialManagement.credentials.createBreadcrumb', {
+        defaultMessage: 'Create credentials',
+      }),
+      href: `/create`,
+    },
+  ];
+}
```
Comment (zengyan-amazon) [inline src/plugins/credential_management/public/components/breadsrumbs.ts:24]: do we plan to have a separate edit page url?

File: src/plugins/credential_management/public/components/breadsrumbs.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import { i18n } from '@osd/i18n';
```
Comment (zengyan-amazon) [inline src/plugins/credential_management/public/components/breadsrumbs.ts:1]: license header

File: src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx [hunk 12]
```diff
@@ -8,3 +8,207 @@
...
+          }}
+          toastLifeTimeMs={6000}
+        />
+      </>
+    );
+  }
+
+  redact = (len: number) => {
+    return '*'.repeat(len);
+  };
```
Comment (zengyan-amazon) [inline src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx:187]: see https://elastic.github.io/eui/#/forms/form-controls#password-field

File: src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx [hunk 13]
```diff
@@ -8,3 +8,207 @@
...
+              password: this.state.password,
+            },
+          }),
+        })
+        .then((res) => {
+          console.log(res);
+        });
+    } catch (e) {
+      console.log(e);
+    }
```
Comment (zengyan-amazon) [inline src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx:209]: show a toast if error

File: src/plugins/credential_management/public/plugin.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (zengyan-amazon) [reply src/plugins/credential_management/public/plugin.ts:76]: I also doubt if legacy path support is necessary

File: src/plugins/credential_management/server/routes/create.ts [hunk 14]
```diff
@@ -14,43 +14,43 @@ import { schema } from '@osd/config-schema';
 import { createHandler } from '../credential';
 
 export function registerCreateRoute(router: IRouter) {
```
Comment (zengyan-amazon) [inline src/plugins/credential_management/server/routes/create.ts:16]: shall we just use saved objects APIs?

Comment (noCharger) [reply src/plugins/credential_management/server/routes/create.ts:16]: > shall we just use saved objects APIs?

Plated it on a separate issue, as discussed offline.

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1850

File: src/plugins/credential_management/common/types.ts (hunk 9 above)
Comment (noCharger) [reply src/plugins/credential_management/common/types.ts:32]: Will finalize it starting from proposal - https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1742

File: src/plugins/credential_management/public/components/breadsrumbs.ts (hunk 10 above)
Comment (noCharger) [reply src/plugins/credential_management/public/components/breadsrumbs.ts:24]: Yes, will included in the editing page PR

File: src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx (hunk 12 above)
Comment (noCharger) [reply src/plugins/credential_management/public/components/create_credential_wizard/create_credential_wizard.tsx:187]: Resolved in #10

File: src/plugins/credential_management/public/components/breadsrumbs.ts (hunk 11 above)
Comment (noCharger) [reply src/plugins/credential_management/public/components/breadsrumbs.ts:1]: Resolved in 10

**Pull request merged**